src/ray/util/ordered_set.h

```diff
@@ -0,0 +1,45 @@
+#ifndef RAY_UTIL_ORDERED_SET_H
+#define RAY_UTIL_ORDERED_SET_H
+
+#include <list>
+#include <unordered_map>
+
+template <typename T>
+class ordered_set {
+ public:
+  ordered_set() {}
+
+  ordered_set(const ordered_set &other) = delete;
+
+  ordered_set &operator=(const ordered_set &other) = delete;
+
+  void push_back(const T &value) {
+    RAY_CHECK(iterators_.find(value) == iterators_.end());
+    auto list_iterator = elements_.insert(elements_.end(), value);
+    iterators_[value] = list_iterator;
+  }
+
+  size_t count(const T &k) const { return iterators_.count(k); }
+
+  void pop_front() {
+    iterators_.erase(elements_.front());
+    elements_.pop_front();
+  }
+
+  const T &front() const { return elements_.front(); }
+
+  size_t size() const noexcept { return iterators_.size(); }
+
+  size_t erase(const T &k) {
+    auto it = iterators_.find(k);
+    RAY_CHECK(it != iterators_.end());
+    elements_.erase(it->second);
+    return iterators_.erase(it);
```

Comment:
The return type is still not right (I'm getting a compile error), the version of iterators_.erase that  taskes an iterator returns an iterator and not a size_t. We could just return 1, since we have `RAY_CHECK(it != iterators_.end());`, but this code isn't generally applicable. It's probably better to return 0 if the element was not found and remove the RAY_CHECK.

Reply:
> the version of iterators_.erase that taskes an iterator returns an iterator and not a size_t. 
That's an overloaded version of the iterator. Currently the user doesn't have a way of calling that because we don't expose iterators to them.

Reply:
@robertnishihara I think he meant that the overload that you are calling (and passing an iterator for) returns an iterator rather than a`size_t`: https://en.cppreference.com/w/cpp/container/unordered_map/erase

Reply:
Whoops, my mistake, thanks!

Reply:
All I'm saying is that with the current version of the code I'm getting this error:
```
In file included from /Users/pcmoritz/ray/src/ray/raylet/scheduling_queue.h:11:
/Users/pcmoritz/ray/src/ray/util/ordered_set.h:37:12: error: no viable conversion from returned value of type 'std::__1::unordered_map<ray::UniqueID,
      std::__1::__list_iterator<ray::UniqueID, void *>, std::__1::hash< ::ray::UniqueID>, std::__1::equal_to<ray::UniqueID>, std::__1::allocator<std::__1::pair<const ray::UniqueID,
      std::__1::__list_iterator<ray::UniqueID, void *> > > >::iterator' (aka '__hash_map_iterator<__hash_iterator<std::__1::__hash_node<std::__1::__hash_value_type<ray::UniqueID,
      std::__1::__list_iterator<ray::UniqueID, void *> >, void *> *> >') to function return type 'size_t' (aka 'unsigned long')
    return iterators_.erase(it);
           ^~~~~~~~~~~~~~~~~~~~
/Users/pcmoritz/ray/src/ray/raylet/scheduling_queue.cc:129:38: note: in instantiation of member function 'ordered_set<ray::UniqueID>::erase' requested here
    tasks_with_resources_[resources].erase(task_id);
```

Reply:
Oh I guess that got resolved thanks @mehrdadn